Git-commit: 317f9f64a9a83386a59ebc8c6f60221a11a7eb2d
From: Zhen Lei <thunder.leizhen@huawei.com>
Date: Mon, 25 Jan 2021 10:08:24 +0800
Subject: [PATCH] perf tools: set the default length of HW_BREAKPOINT_X as 4
 for non X86_64
Patch-mainline: Queued in openEuler repo, version 5.10.0-2.0.0
References: openEuler-21.03
Git-repo: https://gitee.com/openeuler/kernel.git

hulk inclusion
category: bugfix
bugzilla: 47254
CVE: NA

-------------------------------------------------------------------------

Some products may run both 32bit(ILP32) and 64bit APPs on arm64, such as
SD5122. And RTOS will only provides 64bit perf tool. At present, if the
user did not obviously specify the the length of HW_BREAKPOINT_X, perf
will use sizeof(long) as default. For 64bit perf tool, this value is 8.
But the ILP32 syscall only accepts the length 2(Thumb) and 4.

In fact, the original default length is 4 before below patch, because the
length of some instructions is larger than 4 on X86_64.

Use macro HAVE_ARCH_X86_64_SUPPORT to make sure that the default length
is sizeof(long) on X86_64, and HW_BREAKPOINT_LEN_4 for others.

Fixes: aa59a48596d8 ("perf: Don't use 4 bytes as a default instruction breakpoint length")
Signed-off-by: Zhen Lei <thunder.leizhen@huawei.com>
Reviewed-by: Li Bin <huawei.libin@huawei.com>
Signed-off-by: zhangyi (F) <yi.zhang@huawei.com>
Signed-off-by: Li Huafei <lihuafei1@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 tools/perf/util/parse-events.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/tools/perf/util/parse-events.c b/tools/perf/util/parse-events.c
index 3b273580fb84..4e441a3dd487 100644
--- a/tools/perf/util/parse-events.c
+++ b/tools/perf/util/parse-events.c
@@ -955,7 +955,11 @@ int parse_events_add_breakpoint(struct list_head *list, int *idx,
 	/* Provide some defaults if len is not specified */
 	if (!len) {
 		if (attr.bp_type == HW_BREAKPOINT_X)
+#ifdef HAVE_ARCH_X86_64_SUPPORT
 			len = sizeof(long);
+#else
+			len = HW_BREAKPOINT_LEN_4;
+#endif
 		else
 			len = HW_BREAKPOINT_LEN_4;
 	}
-- 
2.31.1

